### PR TITLE
JSUI-2775 Improved accessibility for `ResultPreviewsManager` and `QuerySuggestPreview`

### PR DESCRIPTION
**Pull request by btaillon**
This PR improves accessibility on `ResultPreviewsManager` and `QuerySuggestPreview` and makes it easier to navigate with voiceover. However, there are still many time-consuming ways that this could be improved later (like implementing `aria-keyshortcuts`).

https://coveord.atlassian.net/browse/JSUI-2775

[Deploy](https://dashboard.heroku.com/pipelines/a3535101-5bbf-4a5b-a909-47fcf8c9f149)

### CODE REVIEW

File: src/ui/QuerySuggestPreview/QuerySuggestPreview.ts [hunk 1]
```diff
@@ -178,7 +178,13 @@ export class QuerySuggestPreview extends Component implements IComponentBindings
   }
 
   private buildResultPreview(suggestionText: string, element: HTMLElement, rank: number): ISearchResultPreview {
-    $$(element).addClass('coveo-preview-selectable');
+    element.classList.add('coveo-preview-selectable');
+    const resultLink = element.querySelector(Component.computeSelectorForType(ResultLink.ID)) as HTMLElement;
+    if (resultLink) {
+      element.setAttribute('aria-label', resultLink.textContent);
+      resultLink.setAttribute('role', 'link');
+      resultLink.removeAttribute('aria-level');
```
Comment (olamothe) [inline src/ui/QuerySuggestPreview/QuerySuggestPreview.ts]: Why is this needed ?

File: src/ui/QuerySuggestPreview/QuerySuggestPreview.ts [hunk 2]
```diff
@@ -178,7 +178,13 @@ export class QuerySuggestPreview extends Component implements IComponentBindings
   }
 
   private buildResultPreview(suggestionText: string, element: HTMLElement, rank: number): ISearchResultPreview {
-    $$(element).addClass('coveo-preview-selectable');
+    element.classList.add('coveo-preview-selectable');
+    const resultLink = element.querySelector(Component.computeSelectorForType(ResultLink.ID)) as HTMLElement;
+    if (resultLink) {
+      element.setAttribute('aria-label', resultLink.textContent);
```
Comment (olamothe) [inline src/ui/QuerySuggestPreview/QuerySuggestPreview.ts]: Shouldn't that be directly inside the ResultLink component logic ?

File: src/ui/QuerySuggestPreview/QuerySuggestPreview.ts (hunk 1 above)
Comment (ThibodeauJF) [reply src/ui/QuerySuggestPreview/QuerySuggestPreview.ts]: It seems result link by default has a aria-level of 2 and a role of header if it's the first result link.
@btaillon were those added to one of the result links?

**Comment by ThibodeauJF** [comment]
Some of those can be unit tested too. Accessibility tests are cool but they don't run often in the builds.

**Comment by ThibodeauJF** [comment]
Really smooth to use, good joob!

File: src/ui/QuerySuggestPreview/QuerySuggestPreview.ts (hunk 1 above)
Comment (btaillon) [reply src/ui/QuerySuggestPreview/QuerySuggestPreview.ts]: Ah, I didn't notice that this only affects the first result link. I'll check it out. But yes, I did add this because of the header role and aria-level of 2, which caused Axe to give some warnings.

File: src/ui/QuerySuggestPreview/QuerySuggestPreview.ts (hunk 2 above)
Comment (btaillon) [reply src/ui/QuerySuggestPreview/QuerySuggestPreview.ts]: I'm gonna check it out, but I think `aria-label` might not be allowed when `role` is set to `header`. If it's allowed, I'll add that into `ResultLink`.

**Comment by btaillon** [comment]
> We will also need axe tests inside `./accessibilityTest`

Alright, I'll check that out and see what can be done.

> Some of those can be unit tested too. Accessibility tests are cool but they don't run often in the builds.

I'll check this out too.

File: src/magicbox/ResultPreviewsManager.ts [hunk 3]
```diff
@@ -173,17 +174,25 @@ export class ResultPreviewsManager {
...
       'div',
       {
-        className: 'coveo-preview-container'
+        className: 'coveo-preview-container',
+        id: this.getIdForPreviews()
       },
       (this.resultPreviewsHeader = $$('div', {
-        className: 'coveo-preview-header'
+        className: 'coveo-preview-header',
+        role: 'status'
```
Comment (ThibodeauJF) [inline src/magicbox/ResultPreviewsManager.ts]: curious about what it does add?

Comment (btaillon) [reply src/magicbox/ResultPreviewsManager.ts]: I don't know if there's a particular accessibility use that checks for the `status` role, but it seemed relevant to indicate the header as being a `status` since it only contains text that gets updated whenever the results change.

Comment (ThibodeauJF) [reply src/magicbox/ResultPreviewsManager.ts]: Oh, do you hear it when it changes? with voiceover?

File: unitTests/magicbox/ResultPreviewsManagerTest.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    function buildEnvironment() {
+      return (env = new MockEnvironmentBuilder().build());
+    }
+
+    let container: HTMLElement;
+    function buildContainer() {
+      return buildEnvironment().root.appendChild((container = $$('div').el));
+    }
+
+    let suggestionsListboxElement: HTMLElement;
```
Comment (samisayegh) [inline unitTests/magicbox/ResultPreviewsManagerTest.ts]: I don't think this variable is used anywhere. Could it be removed? It's good to keep the number of `describe`-block-scope variables to a minimum where possible.

File: unitTests/magicbox/ResultPreviewsManagerTest.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+export function ResultPreviewsManagerTest() {
+  describe('ResultPreviewsManager', function() {
+    let env: IMockEnvironment;
+    function buildEnvironment() {
+      return (env = new MockEnvironmentBuilder().build());
+    }
+
+    let container: HTMLElement;
+    function buildContainer() {
+      return buildEnvironment().root.appendChild((container = $$('div').el));
```
Comment (samisayegh) [inline unitTests/magicbox/ResultPreviewsManagerTest.ts]: Rather than having `container`, could we retrieve it by accessing the correct property from the returned environment object?

Similar comment for the other lines in this test file like 26, 33 etc.

File: unitTests/ui/QuerySuggestPreviewTest.ts [hunk 6]
```diff
@@ -61,8 +62,9 @@ export function QuerySuggestPreviewTest() {
       spyOn(Component, 'resolveRoot').and.returnValue(testEnv.root);
     }
 
+    let lastFakeResults: IQueryResults;
     function triggerPopulateSearchResultPreviews(suggestionText: string = 'test', fakeResults?: IQueryResults) {
-      fakeResults = fakeResults || FakeResults.createFakeResults(test.cmp.options.numberOfPreviewResults);
+      lastFakeResults = fakeResults = fakeResults || FakeResults.createFakeResults(test.cmp.options.numberOfPreviewResults);
```
Comment (samisayegh) [inline unitTests/ui/QuerySuggestPreviewTest.ts]: This is not clean. I would either A) force passing a `results` argument, or B) Remove the argument all together. Instead one would need to assign the `results` to a `describe` block scoped variable which would be read by this this function. e.g.

Case A

```
function triggerPopulateSearchResultPreviews(suggestionText: string = 'test', results: IQueryResults) {
    // line 67 is no longer needed
    (test.env.searchEndpoint.search as jasmine.Spy).and.returnValue(Promise.resolve(fakeResults));
    ...
}
```

Case B

```
let results: IQueryResults;

beforeEach(() => {
    results = FakeResults.createFakeResults(test.cmp.options.numberOfPreviewResults)
})

function triggerPopulateSearchResultPreviews(suggestionText: string = 'test') {
    ... read results from variable ...
}

it('some test', () =>{
    results = <something> // change results if necessary
     triggerPopulateSearchResultPreviews()
})
```

File: src/magicbox/ResultPreviewsManager.ts (hunk 3 above)
Comment (btaillon) [reply src/magicbox/ResultPreviewsManager.ts]: Unfortunately I couldn't hear it with the voiceover on MacOS, but I imagine that some other accessibility software might be handling it.

File: accessibilityTest/AccessibilityResultPreviewsManager.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+      getSearchSection().appendChild(querySuggestPreviewElement);
+    }
+
+    function getInput() {
+      return searchBoxElement.getElementsByTagName('input')[0];
+    }
+
+    function triggerEvent(key: KEYBOARD) {
+      const event = new KeyboardEvent('keyup');
+      Object.defineProperty(event, 'keyCode', {
```
[the remark this answers is not in the document]
Comment (btaillon) [reply accessibilityTest/AccessibilityResultPreviewsManager.ts]: I'll try to replace it

File: accessibilityTest/AccessibilityResultPreviewsManager.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+      });
+      Object.defineProperty(event, 'which', {
+        get: () => {
+          return key;
+        }
+      });
+      getInput().dispatchEvent(event);
+    }
+
+    function deferAsync() {
```
[the remark this answers is not in the document]
Comment (btaillon) [reply accessibilityTest/AccessibilityResultPreviewsManager.ts]: This was added as safety to guarantee that every instruction ran after the element is appended has already been run, I'll see if it can be omitted.

File: accessibilityTest/AccessibilityResultPreviewsManager.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+      return new Promise(resolve => setTimeout(resolve));
+    }
+
+    function waitForElement(parent: HTMLElement, condition: (addedElement: HTMLElement) => boolean, subtree = false): Promise<HTMLElement> {
```
[the remark this answers is not in the document]
Comment (btaillon) [reply accessibilityTest/AccessibilityResultPreviewsManager.ts]: Alternatively, we could poll repeatedly with a `querySelector`/`getElementBy...`.
Would you prefer that?

File: accessibilityTest/AccessibilityResultPreviewsManager.ts [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+      await waitForElement(searchBoxElement, element => element.classList.contains('magic-box-suggestion'), true);
+      await deferAsync();
+    }
+
+    async function waitForPreviews() {
+      await waitForElement(searchBoxElement, element => element.classList.contains('coveo-preview-selectable'), true);
+      await deferAsync();
+    }
+
+    async function triggerResultPreviewsManager() {
```
[the remark this answers is not in the document]
Comment (btaillon) [reply accessibilityTest/AccessibilityResultPreviewsManager.ts]: Good idea, I'll rename it.

File: src/magicbox/ResultPreviewsManager.ts [hunk 11]
```diff
@@ -173,17 +174,25 @@ export class ResultPreviewsManager {
...
+    return `coveo-previews-for-${this.lastDisplayedSuggestion.id}`;
+  }
+
   private buildPreviewContainer() {
     return $$(
       'div',
       {
-        className: 'coveo-preview-container'
+        className: 'coveo-preview-container',
+        id: this.getIdForPreviews()
```
[the remark this answers is not in the document]
Comment (btaillon) [reply src/magicbox/ResultPreviewsManager.ts]: Good point, I'll make it into a getter with a more appropriate name.

File: unitTests/magicbox/ResultPreviewsManagerTest.ts [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+import { toArray } from 'lodash';
+
+export function ResultPreviewsManagerTest() {
+  describe('ResultPreviewsManager', function() {
+    let env: IMockEnvironment;
+    function buildEnvironment() {
+      return (env = new MockEnvironmentBuilder().build());
+    }
+
+    let container: HTMLElement;
```
[the remark this answers is not in the document]
Comment (btaillon) [reply unitTests/magicbox/ResultPreviewsManagerTest.ts]: Alright, I'll move those variables

File: unitTests/magicbox/ResultPreviewsManagerTest.ts (hunk 4 above)
Comment (btaillon) [reply unitTests/magicbox/ResultPreviewsManagerTest.ts]: I'll remove it

File: accessibilityTest/AccessibilityResultPreviewsManager.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (samisayegh) [reply accessibilityTest/AccessibilityResultPreviewsManager.ts]: Would it be possible to wait for an amount of time after which we are confident the element is in the page? Ideally this value would not be hard-coded, but rather reference the original delay-causing variable.

File: unitTests/magicbox/ResultPreviewsManagerTest.ts (hunk 5 above)
Comment (btaillon) [reply unitTests/magicbox/ResultPreviewsManagerTest.ts]: Good point, I didn't know `buildEnvironment` already had the `element` property, I'm going to use that instead.

File: unitTests/ui/QuerySuggestPreviewTest.ts (hunk 6 above)
Comment (btaillon) [reply unitTests/ui/QuerySuggestPreviewTest.ts]: Good idea, I'm replacing it with your solution B and introducing a function to create those results that will be called in the `beforeEach` block.

File: accessibilityTest/AccessibilityResultPreviewsManager.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (btaillon) [reply accessibilityTest/AccessibilityResultPreviewsManager.ts]: I'll try to use that instead

File: accessibilityTest/AccessibilityResultPreviewsManager.ts [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+const maxServerResponseTime = 1200;
+const executeQueryDelay = 250;
+
+enum KeyboardKey {
+  UpArrow = 38,
+  DownArrow = 40
+}
+
+function noop() {}
```
Comment (olamothe) [inline accessibilityTest/AccessibilityResultPreviewsManager.ts]: ? ;)

Comment (btaillon) [reply accessibilityTest/AccessibilityResultPreviewsManager.ts]: Woops, removing that function.
